texar/core/attention_mechanism.py

```diff
@@ -110,9 +112,36 @@ def __init__(self,
             score_mask_value = torch.tensor(-np.inf)
         self.score_mask_value = score_mask_value
 
+        self._encoder_output_size = encoder_output_size
+
+        self._query: Optional[torch.Tensor] = None
         self._values: Optional[torch.Tensor] = None
         self._keys: Optional[torch.Tensor] = None
-        self._encoder_output_size: Optional[int] = None
+
+    def forward(self,  # type: ignore
+                query: torch.Tensor,
+                memory: torch.Tensor,
+                memory_sequence_length: Optional[torch.Tensor] = None):
+        r"""Preprocess the memory and query.
+
+        Args:
+            query: tensor, shaped ``[batch_size, query_depth]``.
+            memory: the memory to query; usually the output of an RNN encoder.
+                This tensor should be shaped ``[batch_size, max_time, ...]``.
+            memory_sequence_length (optional): sequence lengths for the batch
+                entries in memory.  If provided, the memory tensor rows are
+                masked with zeros for values past the respective sequence
+                lengths.
+        """
+        self._query = self.query_layer(query) if self.query_layer else query
```

Comment:
Is `self._query` cleared by the end of `forward`? This should also be cleared.

Actually, I would prefer making `query` the return value of the base class `forward` method, because we don't need to cache it, and storing it as an attribute gives more confusing `Optional` type annotations.